aclk/aclk_query.c

```diff
@@ -551,25 +551,37 @@ static int aclk_process_query(struct aclk_query_thread *t_info)
 
     switch (this_query->cmd) {
         case ACLK_CMD_ONCONNECT:
-            debug(D_ACLK, "EXECUTING on connect metadata command");
-            ACLK_SHARED_STATE_LOCK;
-            meta_state = aclk_shared_state.metadata_submitted;
-            aclk_shared_state.metadata_submitted = ACLK_METADATA_SENT;
-            ACLK_SHARED_STATE_UNLOCK;
-            aclk_send_metadata(meta_state);
+            if (!host)
+                fatal("ACLK_CMD_ONCONNECT needs host pointer");
+
+            debug(D_ACLK, "EXECUTING on connect metadata command for host \"%s\" GUID \"%s\"",
+                host->hostname,
+                host->machine_guid);
+
+            rrdhost_aclk_state_lock(host);
+            meta_state = host->aclk_state.metadata;
+            host->aclk_state.metadata = ACLK_METADATA_SENT;
+            rrdhost_aclk_state_unlock(host);
+            aclk_send_metadata(meta_state, host);
             break;
 
         case ACLK_CMD_CHART:
+            if (!host)
+                fatal("ACLK_CMD_CHART needs host pointer");
+
```

Comment:
You check for (!host) twice (a few lines below that)

- consider using unlikely
- How can we remove the fatal? When will that happen, can we ignore and continue?

Reply:
hmmm, well if we don't have host here it is a logical programming error, not a runtime error. We could continue though and just ignore the message?

Reply:
I would say log the error / ignore that message and continue instead of fatal

Reply:
changed